### PR TITLE
mon: show io status quickly if no update in a long period

### PR DESCRIPTION
**Pull request by LiumxNL**
Signed-off-by: Mingxin Liu <mingxin@xsky.com>

### CODE REVIEW

**Comment by LiumxNL** [comment]
retest this please

File: src/mon/PGMap.cc [hunk 1]
```diff
@@ -1477,7 +1477,11 @@ void PGMap::update_delta(CephContext *cct,
   delta_t -= *last_ts;  // take the last timestamp we saw
   *last_ts = ts;        // @p ts becomes the last timestamp we saw
 
-  // calculate a delta, and average over the last 2 deltas.
+  // adjust delta_t, quick start if there is no update in a long period
+  if (*result_ts_delta == utime_t())
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: yeah, it's hard to be accurate in this case.

Comment (tchaikov) [inline src/mon/PGMap.cc]: better off putting
```c++
if (result_ts_delta->is_zero()) {
  delta_t = std::min(delta_t, utime_t(2 * g_conf->mon_delta_reset_interval, 0));
}
```

**Comment by LiumxNL** [comment]
@tchaikov done with your suggestion.

**Comment by tchaikov** [comment]
```
85/166 Test  #92: check-generated.sh ......................***Failed   17.74 sec
Enivronment Variables Already Set
checking ceph-dencoder generated test instances...
numgen type
3 ACLGrant
2 ACLGranteeType
2 ACLOwner
2 ACLPermission
3 AuthMonitor::Incremental
2 BitVector<2>
2 BloomHitSet
2 Capability
24395: copy ctor not supported
24395: copy operator= not supported
24395: copy operator= not supported
24395: copy ctor not supported
2 CompatSet
1 CrushWrapper
2 DBObjectMap::State
2 DBObjectMap::_Header
24395: copy operator= not supported
24395: copy ctor not supported
2 DecayCounter
2 ECSubRead
2 ECSubReadReply
3 ECSubWrite
24395: copy ctor not supported
24395: copy operator= not supported
2 ECSubWriteReply
2 ECUtil::HashInfo
24395: copy ctor not supported
24395: copy operator= not supported
24395: copy ctor not supported
24395: copy operator= not supported
2 ECommitted
1 EExport
2 EFragment
2 EImportFinish
1 EImportStart
1 EMetaBlob
1 EMetaBlob::dirlump
1 EMetaBlob::fullbit
2 EMetaBlob::nullbit
1 EMetaBlob::remotebit
2 EOpen
24395: copy ctor not supported
24395: copy operator= not supported
1 EResetJournal
1 ESession
1 ESessions
1 ESlaveUpdate
1 ESubtreeMap
1 ETableClient
1 ETableServer
1 EUpdate
2 ExplicitHashHitSet
  sorting json output for nondeterministic object
  sorting json output for nondeterministic object
2 ExplicitObjectHitSet
  sorting json output for nondeterministic object
  sorting json output for nondeterministic object
1 FSMap
4 HitSet
  sorting json output for nondeterministic object
  sorting json output for nondeterministic object
  sorting json output for nondeterministic object
  sorting json output for nondeterministic object
8 HitSet::Params
1 InoTable
1 InodeStore
2 JournalPointer
3 Journaler::Header
2 LevelDBStoreStats
1 LogEntry
2 LogEntryKey
1 LogSummary
0 MAuth
0 MAuthReply
0 MCacheExpire
0 MClientCapRelease
0 MClientCaps
0 MClientLease
0 MClientReconnect
0 MClientReply
0 MClientRequest
0 MClientRequestForward
0 MClientSession
0 MClientSnap
0 MCommand
0 MCommandReply
3 MDSCacheObjectInfo
1 MDSMap
2 MDSMap::mds_info_t
0 MDentryLink
0 MDentryUnlink
0 MDirUpdate
0 MDiscover
0 MDiscoverReply
0 MExportCaps
0 MExportCapsAck
0 MExportDir
0 MExportDirAck
0 MExportDirCancel
0 MExportDirDiscover
0 MExportDirDiscoverAck
0 MExportDirFinish
0 MExportDirNotify
0 MExportDirNotifyAck
0 MExportDirPrep
0 MExportDirPrepAck
0 MForward
0 MGetPoolStats
0 MGetPoolStatsReply
0 MHeartbeat
0 MInodeFileCaps
0 MLock
0 MLog
0 MLogAck
0 MMDSBeacon
0 MMDSCacheRejoin
0 MMDSFindIno
0 MMDSFindInoReply
0 MMDSFragmentNotify
0 MMDSLoadTargets
0 MMDSMap
0 MMDSResolve
0 MMDSResolveAck
0 MMDSSlaveRequest
0 MMDSTableRequest
0 MMonCommand
0 MMonCommandAck
0 MMonElection
0 MMonGetMap
0 MMonGetVersion
0 MMonGetVersionReply
0 MMonGlobalID
0 MMonJoin
0 MMonMap
0 MMonPaxos
0 MMonProbe
0 MMonScrub
0 MMonSubscribe
0 MMonSubscribeAck
0 MMonSync
0 MOSDAlive
0 MOSDBoot
0 MOSDFailure
0 MOSDMap
0 MOSDOp
0 MOSDOpReply
0 MOSDPGBackfill
0 MOSDPGCreate
0 MOSDPGInfo
0 MOSDPGLog
0 MOSDPGNotify
0 MOSDPGQuery
0 MOSDPGRemove
0 MOSDPGScan
0 MOSDPGTemp
0 MOSDPGTrim
0 MOSDPing
0 MOSDRepScrub
0 MOSDScrub
0 MOSDSubOp
0 MOSDSubOpReply
0 MPGStats
0 MPGStatsAck
0 MPing
0 MPoolOp
0 MPoolOpReply
0 MRemoveSnaps
0 MRoute
0 MStatfs
0 MStatfsReply
0 MWatchNotify
1 MgrMap
8 MonCap
3 MonMap
1 MonitorDBStore::Op
2 MonitorDBStore::Transaction
2 OSDMap
1 OSDMap::Incremental
3 OSDSuperblock
24395: copy ctor not supported
24395: copy operator= not supported
24395: copy operator= not supported
24395: copy ctor not supported
2 ObjectCacheInfo
2 ObjectMetaInfo
1 ObjectRecoveryInfo
2 ObjectRecoveryProgress
4 ObjectStore::Transaction

The following tests FAILED:
	 92 - check-generated.sh (Failed)
```

see https://jenkins.ceph.com/job/ceph-pull-requests/22052/console

retest this please.

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
@LiumxNL could you rebase you change against master for a green jenkins run?

**Comment by tchaikov** [comment]
retest this please.

File: src/mon/PGMap.cc [hunk 2]
```diff
@@ -1514,7 +1514,12 @@ void PGMap::update_delta(CephContext *cct,
   delta_t -= *last_ts;  // take the last timestamp we saw
   *last_ts = ts;        // @p ts becomes the last timestamp we saw
 
-  // calculate a delta, and average over the last 2 deltas.
+  // adjust delta_t, quick start if there is no update in a long period
+  if (result_ts_delta->is_zero()) {
```
Comment (liewegas) [inline src/mon/PGMap.cc]: Why is this condition needed?  It seems like you could do the min unconditionally...

Comment (LiumxNL) [inline src/mon/PGMap.cc]: you are quite right, ty

**Comment by LiumxNL** [comment]
@tchaikov sorry for  the latency, jenkins is ok now.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-23_19:52:01-rados-wip-sage-testing2---basic-smithi/

**Pull request merged**